[EASI-3023] - Allow storing feedback on governance requests in IT Gov V2

DylanSpOddball opened the pull request:
# EASI-3023

## Changes and Description

- Added a new "Governance Request Feedback" entity that represents feedback on governance requests in IT Gov V2. (Note: the name is meant to distinguish it from TRB Request Feedback and GRT Feedback)
  - SQL table
  - Go model
  - Go `storage` methods
    - `CreateGovernanceRequestFeedback()` to save a governance request feedback item to the database
    - `GetGovernanceRequestFeedbacksByIntakeID()` to get all governance request feedback items for a given system intake from the database.
  - GraphQL type (also added as a child of the `SystemIntake` GraphQL type)
  - GraphQL query to get all governance request feedback items for a given intake's ID.
  - Implementation of the necessary GraphQL resolver.
  - Validation logic to make sure that the feedback's source action and target form are compatible; a "Progress to New Step" action should always have a `NO_TARGET_PROVIDED` target, while a "Request Edits" action should always have a target of `INTAKE_REQUEST`, `DRAFT_BUSINESS_CASE`, or `FINAL_BUSINESS_CASE`.
- Added code to `cmd/devdata/main.go` to save example governance request feedback items as part of our seed data.

## How to test this change

### Testing happy path

1. Bring up the app backend with `scripts/dev up:backend`.
2. Run `db:seed`. This will test `CreateGovernanceRequestFeedback()` by calling it to save feedback items as part of the seed data.
3. Run the following GraphQL query in the GQL Playground, with these four intake IDs (hardcoded in the seed data):
  a. `4d3f9821-e043-42bf-9cd0-faa5f053ed32`
  b. `29486f85-1aba-4eaf-a7dd-6137b9873adc`
  c. `ce874e71-de26-46da-bbfe-a8e3af960108`
  d. `67eebec8-9242-4f2c-b337-f674686a5ab5`
```gql
query {
  systemIntake(id:"<intake ID from above>") {
    requestName
    governanceRequestFeedbacks {
      id
      intakeId
      feedback
      sourceAction
      targetForm
      createdBy
      createdAt
      modifiedBy
      modifiedAt
    }
  }
}
```
These should return the seeded data. The first intake ID should have one feedback item with feedback on the "progress to new step" action, with a `sourceAction` of `PROGRESS_TO_NEW_STEP` and a `targetForm` of `NO_TARGET_PROVIDED`. The three other intakes should each have a feedback item with feedback on the "edits requested" action, all with `sourceAction` set to `REQUEST_EDITS`; `targetForm` will vary, with one result each for feedback on the intake request, on the draft business case, and on the final business case. The feedback `id` field should be dynamic and have a different value from the intake ID each time.

### Testing validation logic

1. With the app backend up, run `scripts/dev db:clean`.
2. Edit the seed data in `cmd/devdata/main.go` - in the `makeSystemIntakeWithEditsRequested()` function, in the definition of `feedback`, change `SourceAction: models.GovernanceRequestFeedbackSourceActionRequestEdits` to `SourceAction: models.GovernanceRequestFeedbackSourceActionProgressToNewStep`.
3. Run `scripts/dev db:seed`. You should get an error, the final panic message should be `pq: insert or update on table "governance_request_feedback" violates foreign key constraint "governance_request_feedback_source_action_fkey"`.

## Notes
- The author of the feedback is represented by the usual `created_by` metadata.
- The team of the feedback author isn't currently represented at any level; it's assumed to be the governance team, which the frontend will display. (See [Figma](https://www.figma.com/file/ChzAP34A2DVvQUNQwD7lCt/IT-Governance-Next?type=design&node-id=428-104044&t=Ayzi4xxIsbWFMaTN-0))
- While this PR includes a `Store` method for saving a `GovernanceRequestFeedback` to the database, there's no GQL mutation for creating feedback. This is intentional; EASI-2888 and EASI-3019 will implement the actions that can add feedback, those tickets will add GQL mutations that include feedback to save. 
- Source/target validation logic
  - I originally experimented with implementing the source/target validation logic in SQL, as you can see in https://github.com/CMSgov/easi-app/pull/2067/commits/7b9f1b1f258c574ab280725d9dd54b6057a8939f. While it works, it would be annoying to maintain and modify; every migration modifying it would have to redefine it, starting by copy-pasting the most recent definition.
  - I then tried implementing the validation logic in Go; see https://github.com/CMSgov/easi-app/pull/2067/commits/5bdccf4d0f4cc1ac7045f75d3161ac51ca3f5b27. This worked, but I was thinking about how best to enforce the validation logic, without allowing Go code to bypass it.
  - I ultimately figured out another SQL-based solution, using the `feedback_valid_source_target_combinations` table. This should be easier to maintain; adding or removing valid source/target combinations can be done simply by inserting/deleting rows in this table. Having the validation logic be implemented in SQL means that it'll be _always_ enforced for data we save.
- Keeping the various enums (in SQL, Go, and GraphQL) in sync across languages could get annoying and hard to keep consistent.
- I'm not sure if there is, or will be, an enum for all possible actions in IT Gov V2, in any of SQL, Go, or GraphQL. If there is, we should potentially put a single definition in a centralized file and use it everywhere, rather than duplicating the list of actions in multiple places. (The same might also apply to an enum with all forms)

## Questions for the team
- ~~The GQL type doesn't include the system intake's ID; should it? I noticed that most TRB-related GQL types, such as `TRBRequestFeedback`, contain a `trbRequestId: UUID!` field pointing back to their "parent" TRB request; should I add that to the `GovernanceRequestFeedback` GQL type in this PR?~~ I ended up adding an `intakeId` field to the GQL schema.
- ~~Is it worth having the validation logic to make sure that the `sourceAction`/`targetForm` combination matches what we expect? If so, is calling it from the storage method for creating feedback the right place? We could potentially add a pseudo-constructor function in `pkg/models/governance_request_feedback.go` that validates instead, though we don't usually do that with our models.~~ See notes above on the approach I took for validation.
- The names for our constants in Go enums, i.e. `GovernanceRequestFeedbackTargetFinalBusinessCase`, can get pretty long, which can lead to mistaking one value for another; should we change the naming convention?

StevenWadeOddball commented on pkg/models/governance_request_feedback.go:
```diff
@@ -0,0 +1,32 @@
+type GovernanceRequestFeedback struct {
+	BaseStruct
+	IntakeID     uuid.UUID `db:"intake_id"`
+	Feedback     string
```
I noticed that these fields don't all have struct tags (and they only have them for db and not json as well). Am I guessing correctly that sqlx a GQL are able to automatically convert the name if the name is one word?

 In `MINT` every struct has both tags provided on every field, but it looks like this is consistent with other locations in `EASi`

DylanSpOddball replied:
Yeah, they both have some defaults, which we generally rely on in EASi:
- sqlx just calls `strings.ToLower()` on the field name. See http://jmoiron.github.io/sqlx/, "Controlling Name Mapping"
- I'm not sure if gqlgen configures its own defaults or if they're set up in the `encoding/json` package from the Go stdlib, but they're definitely present.

We're not entirely consistent about it, though, and I wouldn't be against including both tags as regular practice, especially if that's what MINT is doing as well.

StevenWadeOddball replied:
Awesome, thanks Dylan! I like including both struct tags myself. If you like it too, I think it could be a nice to adopt as a regular practice for EASi (I'll leave it up to you if you want to implement in this PR)

DylanSpOddball replied:
Added `json` and `db` struct tags for all fields in https://github.com/CMSgov/easi-app/pull/2067/commits/50cae8fdec74bc25f960bd40507e015def41943c, since I was making edits anyways.

StevenWadeOddball commented on pkg/storage/governance_request_feedback.go:
```diff
@@ -0,0 +1,129 @@
+	return feedbacks, nil
+}
+
+func validateSourceActionTargetFormCombination(feedback models.GovernanceRequestFeedback) bool {
```
This function seems to work well, but I wonder if it would make more sense to live with the definition of the struct itself? Perhaps as a receiver function in the models package? I think of the storage package specifically being responsible for reading and writing to that database. What do you think?

DylanSpOddball replied:
No longer applicable due to enforcing validation in SQL.

StevenWadeOddball commented on pkg/storage/governance_request_feedback.go:
```diff
@@ -0,0 +1,129 @@
+func (s *Store) GetGovernanceRequestFeedbacksByIntakeID(ctx context.Context, intakeID uuid.UUID) ([]*models.GovernanceRequestFeedback, error) {
+	feedbacks := []*models.GovernanceRequestFeedback{}
+
+	stmt, err := s.db.PrepareNamed(`SELECT * FROM governance_request_feedback WHERE intake_id = :intake_id`)
```
What do you think about explicitly listing out the columns for the `SELECT` instead of `SELLECT *`?  We've have a couple issues where `SELECT *` creates a transient error in the app during deployment (because the database is updated before the app code is updated). This paradigm certainly has precedence in the `EASi` code base if you prefer it.

DylanSpOddball replied:
I do want to think about this in general; I still need to read over what you posted in Slack about this discussion. I think I'm going to stay with `SELECT *` on this particular PR to just keep things moving, but I'm definitely open to revisiting our approach.

StevenWadeOddball replied:
Thanks Dylan! Yeah that sounds perfect to me.

StevenWadeOddball commented on go.mod:
```diff
@@ -96,7 +96,7 @@ require (
 	go.uber.org/multierr v1.8.0 // indirect
 	golang.org/x/crypto v0.9.0 // indirect
-	golang.org/x/exp v0.0.0-20220303212507-bbda1eaf7a17 // indirect
+	golang.org/x/exp v0.0.0-20220303212507-bbda1eaf7a17
```
Do you know why this changed from an indirect reference?

DylanSpOddball replied:
I think it's because `pkg/storage/governance_request_feedback.go` uses `golang.org/x/exp/slices`.

StevenWadeOddball replied:
Thanks Dylan!

StevenWadeOddball commented on pkg/storage/governance_request_feedback_test.go:
If you agree that the validation logic could live in models, I'd suggest this test file could be moved as well. I like your method here to pass an array and test every case!

DylanSpOddball replied:
No longer applicable due to enforcing validation in SQL.

StevenWadeOddball reviewed (approved):
Thanks @DylanSpOddball, this looks good an I think it will address all the needs the ticket presents ( as well as the other use cases you have looked ahead to). I pulled the changes locally and tested via the script you provided. 

I provided some optional feedback you can choose to implement or not, but this seems to work well, so it doesn't need to block deployment if you prefer.

Also to give my thoughts for some of your questions.
-  I probably would include the systemIntakeID in the GQL type, since we already return it from SQL. It would be convenient to have it incase it is needed, or is needed for any other future resolver. However I don't think that it's absolutely required in the current resolver.

- I do like the idea of changing the package the validation logic is in. I can see why it could be in storage, but I think the models package probably makes more sense if we want to handle it in the app code. FWIW, I liked the approach of putting the validation in SQL as compared to the app code. I like that the constraint always enforces the check, vs now where it could potentially be bypassed. I think either way works though.

- 🤔 for the enum names, I'm personally fine with abbreviating the parent type name.  I generally create an acronym, and then add the value. I think listing the whole name is fine too if you prefer.

DylanSpOddball commented:
Updated in response to Steven's comments and some thinking about how best to enforce validation; I went back to using a SQL-based approach instead of validating in Go code, but this should be easily maintainable. I also incorporated a couple of other minor improvements from feedback while I was making changes.
